infer/default.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infer
+
+import (
+	"fmt"
+	"slices"
+
+	provider "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/middleware/schema"
+)
+
+// Default creates an inferred provider which applies as many defaults as possible.
+func Default() DefaultProvider {
+	return DefaultProvider{}
+}
+
+type DefaultProvider struct {
+	name, version string
+	components    []InferredComponent
+}
```

Comment:
If we're going to add a builder to `infer` we should make sure it's compatible with the custom resource APIs and that it offers a reasonably smooth progression from the easy defaults to highly customized.